### ISSUE TITLE
Null exception when using "Save as Umbraco content node" on approval

### ISSUE THREAD

**Issue by hjaltedaniel-umbraco** [umbraco/Umbraco.Forms.Issues#374]
When setting up post moderation for a form, and use the "Save as Umbraco content node" workflow, you get a null exception when trying to approve the entry.

Reproduction
------------
### Bug summary
When "Save as Umbraco content node" is done on approval, you get a null exception:
`System.NullReferenceException: Object reference not set to an instance of an object.
   at Umbraco.Forms.Core.Services.PageService.GetPageElements()
   at Umbraco.Forms.Core.Services.WorkflowService.ExecuteWorkflows(List`1 workflows, RecordEventArgs e)
   at Umbraco.Forms.Core.Services.WorkflowService.ExecuteWorkflows(Record record, Form form, FormState state, Boolean editMode)
   at Umbraco.Forms.Core.Services.RecordService.Approve(Record record, Form form)
   at Umbraco.Forms.Core.Providers.RecordActions.Recordsets.ApproveRecordSet.Execute(List`1 records, Form form)
   at Umbraco.Forms.Web.Editors.RecordController.PostExecuteRecordSetAction(RecordActionExecution model)
   at lambda_method(Closure , Object , Object[] )
   at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ActionExecutor.<>c__DisplayClass6_2.<GetExecutor>b__2(Object instance, Object[] methodParameters)
   at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ActionExecutor.Execute(Object instance, Object[] arguments)
   at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ExecuteAsync(HttpControllerContext controllerContext, IDictionary`2 arguments, CancellationToken cancellationToken)`

### Specifics
This issue is seen on these versions:
CMS: 8.6.3
Forms: 8.4.1

### Steps to reproduce

1. Create a form and enable post moderation.
2. Set "Save as Umbraco content node" to be the workflow on approval
3. Submit an entry
4. Try to approve it - you'll get the exception above.

### Expected result

The content of the form is saved as a content node.

### Actual result

Exception.

---
_This item has been added to our backlog [AB#8948](https://dev.azure.com/umbraco/243e7927-03b2-44e2-908f-d4ac7ea5daaa/_workitems/edit/8948)_

**Comment by robozum** [comment]
We encountered this bug whilst working on a project and our client has enquired about estimates on when this bug will be investigated as it's affecting functionality on the project? Thanks.

**Comment by gabador-mof** [comment]
I get the same error when trying to submit a form using the IRecordService. When I try to submit the entry, I get this exception

**Comment by chrisdent14** [comment]
I have been attempted to save a record to an Umbraco Forms form using the API but I am getting an Object reference error as   every time.

```
System.NullReferenceException: Object reference not set to an instance of an object.
   at Umbraco.Forms.Core.Services.PageService.GetPageElements()
   at Umbraco.Forms.Core.Services.WorkflowService.ExecuteWorkflows(List`1 workflows, RecordEventArgs e)
   at Umbraco.Forms.Core.Services.WorkflowService.ExecuteWorkflows(Record record, Form form, FormState state, Boolean editMode)
   at Umbraco.Forms.Core.Services.RecordService.Submit(Record record, Form form)
```

Umbraco: 8.6.4
Umbraco Forms: 8.5.3

Umbraco forms are being saved in the database.

I am using the IFormService and IRecordService and the error occurs when I call
`recordService.Submit(record, fileDownloadsForm);`

The whole method is below for reference. 

I have also contacted Umbraco support but thought it would be worth dropping this here in case anyone else can help.
I will update this if I hear anything from support around this.


```
private void SaveFormData(FileDownloadFormViewModel model)
{
	try
	{
		if (formService.Get("File downloads") is Form fileDownloadsForm)
		{
			var siteRoot = cmsService.GetSiteRoot(model.CurrentNodeKey);
			var siteName = string.Concat(siteRoot.Ancestor<SiteRegion>().Name, " - ", siteRoot.Ancestor<SiteContainer>().Name);

			var nameField = fileDownloadsForm.AllFields.FirstOrDefault(f => f.Alias == "name");
			var companyField = fileDownloadsForm.AllFields.FirstOrDefault(f => f.Alias == "company");
			var emailField = fileDownloadsForm.AllFields.FirstOrDefault(f => f.Alias == "emailAddress");
			var siteField = fileDownloadsForm.AllFields.FirstOrDefault(f => f.Alias == "site");

			var record =
				new Record()
				{
					Created = DateTime.Now,
					Form = fileDownloadsForm.Id,
					State = FormState.Approved,
					RecordFields =
						new Dictionary<Guid, RecordField>
						{
							{ nameField.Id, new RecordField(nameField) { Values = new List<object> { model.Name } } },
							{ companyField.Id, new RecordField(companyField) { Values = new List<object> { model.Company } } },
							{ emailField.Id, new RecordField(emailField) { Values = new List<object> { model.Email } } },
							{ siteField.Id, new RecordField(siteField) { Values = new List<object> { siteName } } },
						}
				};

			record.RecordData = record.GenerateRecordDataAsJson();

			recordService.Submit(record, fileDownloadsForm);
		}
	}
	catch (Exception ex)
	{
		logger.Error(typeof(FileDownloadController), ex);
	}
}
```

**Comment by Matthew-Wise** [comment]
I have recreated this issue on 
Umbraco 8.7
Umbraco forms 85.3

Response from the server:

```

)]}',
{"Message":"An error has occurred.","ExceptionMessage":"Object reference not set to an instance of an object.","ExceptionType":"System.NullReferenceException","StackTrace":"   at Umbraco.Forms.Core.Services.PageService.GetPageElements()\r\n   at Umbraco.Forms.Core.Services.WorkflowService.ExecuteWorkflows(List`1 workflows, RecordEventArgs e)\r\n   at Umbraco.Forms.Core.Services.WorkflowService.ExecuteWorkflows(Record record, Form form, FormState state, Boolean editMode)\r\n   at Umbraco.Forms.Core.Services.RecordService.Approve(Record record, Form form)\r\n   at Umbraco.Forms.Core.Providers.RecordActions.Recordsets.ApproveRecordSet.Execute(List`1 records, Form form)\r\n   at Umbraco.Forms.Web.Editors.RecordController.PostExecuteRecordSetAction(RecordActionExecution model)\r\n   at lambda_method(Closure , Object , Object[] )\r\n   at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ActionExecutor.<>c__DisplayClass6_2.<GetExecutor>b__2(Object instance, Object[] methodParameters)\r\n   at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ActionExecutor.Execute(Object instance, Object[] arguments)\r\n   at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ExecuteAsync(HttpControllerContext controllerContext, IDictionary`2 arguments, CancellationToken cancellationToken)\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Web.Http.Controllers.ApiControllerActionInvoker.<InvokeActionAsyncCore>d__1.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__6.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__6.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Web.Http.Filters.ActionFilterAttribute.<ExecuteActionFilterAsyncCore>d__5.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__6.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__6.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Web.Http.Filters.ActionFilterAttribute.<ExecuteActionFilterAsyncCore>d__5.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__6.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__6.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Web.Http.Filters.ActionFilterAttribute.<ExecuteActionFilterAsyncCore>d__5.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__6.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__6.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Web.Http.Filters.ActionFilterAttribute.<ExecuteActionFilterAsyncCore>d__5.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__6.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Web.Http.Filters.ActionFilterAttribute.<CallOnActionExecutedAsync>d__6.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Web.Http.Filters.ActionFilterAttribute.<ExecuteActionFilterAsyncCore>d__5.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Web.Http.Controllers.ActionFilterResult.<ExecuteAsync>d__5.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Web.Http.Filters.AuthorizationFilterAttribute.<ExecuteAuthorizationFilterAsyncCore>d__3.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Web.Http.Filters.AuthorizationFilterAttribute.<ExecuteAuthorizationFilterAsyncCore>d__3.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Web.Http.Filters.AuthorizationFilterAttribute.<ExecuteAuthorizationFilterAsyncCore>d__3.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Web.Http.Filters.AuthorizationFilterAttribute.<ExecuteAuthorizationFilterAsyncCore>d__3.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Web.Http.Controllers.ExceptionFilterResult.<ExecuteAsync>d__6.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Web.Http.Controllers.ExceptionFilterResult.<ExecuteAsync>d__6.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Web.Http.Dispatcher.HttpControllerDispatcher.<SendAsync>d__15.MoveNext()"}
```

Looks like it is throwing a null reference exception due to PageService GetPageElements wanting to the current page Id from the request.

Had to remove the umbraco folder to upload but here is a clean install with the issue present:

[UmbracoFormsTest.zip](https://github.com/umbraco/Umbraco.Forms.Issues/files/5310846/UmbracoFormsTest.zip)

**Comment by jaandrews** [comment]
I get this same error when I add a custom workflow in umbraco 8.6.5. Like **chrisdent14**, I have umbraco forms configured to store it's data in the database and encounter the same error when trying to approve a submission.

**Comment by nul800sebastiaan** [comment]
Fixed for the upcoming Forms v8.5.4

**Comment by chrisdent14** [comment]
@nul800sebastiaan Can confirm the update to 8.5.4 has resolved the issue for me.

To confirm my issue was that I could not add a record to a form (stored in the DB) via the API. v8.5.3 was not working, but upgraded to v8.5.4 and this has resolved the issue however I needed to ensure I had set the 'UmbracoPageId' property of the Record object otherwise I got the following exception:

`"   at Umbraco.Web.PublishedContentExtensions.GetCultureFromDomains(IPublishedContent content, Uri current) in D:\\a\\1\\s\\src\\Umbraco.Web\\PublishedContentExtensions.cs:line 188"`
